design/topology_proposal.md

```diff
@@ -0,0 +1,141 @@
+# Topology aware scheduling proposal
+
+## Background
+
+There is often interest in making the scheduler aware of factors such as
+availability zones. This document proposes a generic way to customize scheduler
+behavior based on labels attached to nodes.
+
+## Approach
+
+Define a repeated field under `Placement` that the scheduler consults when it
+places tasks. These "placement preferences" would be listed in decreasing order
+of precedence, and would have higher precedence than the default scheduler
+logic.
+
+These placement preferences would be interpreted based on their types, but
+the initially supported "spread over" message would tell the scheduler to spread
+tasks evenly between nodes which have each distinct value of the referenced node
+or engine label.
+
+## Protobuf changes
+
+In the `Placement` message that exists in `TaskSpec`, add a repeated field
+called `Preferences`. (Note: I don't think `Preferences` is a very good name.
+Looking for suggestions.)
+
+```
+repeated PlacementPreference preferences = 2;
+```
+
+`PlacementPreference` will be a message that specifies how to act on a label.
+The initially supported preference would be "spread" and "pack".
+
+```
+message SpreadOver {
+	string descriptor = 1; // label descriptor, such as engine.labels.az
+	// TODO: support node information beyond engine and node labels
+
+	// TODO: in the future, add a map that provides weights for weighted
+	// spreading.
+}
+
+message PlacementPreference {
+	oneof Preference {
+                SpreadOver spread = 1;
+        }
+
+	Preference pref = 1;
+}
+```
+
+## Behavior
+
+A simple use of this feature would be to spread tasks evenly between multiple
+availability zones. The way to do this would be to create an engine label on
+each node indicating its availability zone, and then create a
+`PlacementPreference` with type `SpreadOver` which references the engine label.
+The scheduler would prioritize balance between the availability zones, and if
+it ever has a choice between multiple nodes in the preferred availability zone
+(or a tie between AZs), it would choose the node based on its built-in logic.
+As of 1.13, this logic will prefer to schedule a task on the node which has the
+fewest tasks associated with the particular service.
+
+A slightly more complicated use case involves hierarchical topology. Say there
+are two datacenters, which each have four rows, each row having 20 racks. To
+spread tasks evenly at each of these levels, there could be three `SpreadOver`
+messages in `Preferences`. The first would spread over datacenters, the second
```

Comment:
What does the math look like here? Would it minimize items per bucket then fallback to a compounding factor? When would this carryover be triggered?